pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/ModelInterpreter.groovy

```diff
@@ -194,69 +212,100 @@ class ModelInterpreter implements Serializable {
+                        if (asStage?.parallelContent) {
+                            if (evaluateWhen(stageOrGroup.when)) {
+                                withCredentialsBlock(stageOrGroup.environment) {
+                                    withEnvBlock(stageOrGroup.getEnvVars(script)) {
+                                        script.parallel(getParallelStages(root, parentAgent, asStage, firstError, false, false, false))
                                     }
                                 }
                             } else {
-                                Utils.logToTaskListener("Stage '${thisStage.name}' skipped due to when conditional")
-                                Utils.markStageSkippedForConditional(thisStage.name)
-                                script.parallel(getParallelStages(root, parentAgent, thisStage, firstError, parentStage, false, false, true))
+                                Utils.logToTaskListener("${thisType} '${stageOrGroup.name}' skipped due to when conditional")
+                                Utils.markStageSkippedForConditional(stageOrGroup.name)
+                                isSkipped = true
+                                script.parallel(getParallelStages(root, parentAgent, asStage, firstError, false, false, true))
                             }
                         } else {
-                            inDeclarativeAgent(thisStage, root, thisStage.agent) {
-                                if (evaluateWhen(thisStage.when)) {
-                                    withCredentialsBlock(thisStage.environment) {
-                                        withEnvBlock(thisStage.getEnvVars(script)) {
-                                            toolsBlock(thisStage.agent ?: root.agent, thisStage.tools, root) {
-                                                // Execute the actual stage and potential post-stage actions
-                                                executeSingleStage(root, thisStage, parentAgent)
+                            inDeclarativeAgent(stageOrGroup, root, stageOrGroup.agent) {
+                                if (evaluateWhen(stageOrGroup.when)) {
+                                    withCredentialsBlock(stageOrGroup.environment) {
+                                        withEnvBlock(stageOrGroup.getEnvVars(script)) {
+                                            toolsBlock(stageOrGroup.tools, stageOrGroup.agent ?: root.agent, parent?.tools ?: root.tools) {
+                                                if (asStage) {
+                                                    // Execute the actual stage and potential post-stage actions
+                                                    executeSingleStage(root, asStage, parentAgent)
+                                                } else if (asGroup) {
+                                                    evaluateSequentialStages(root, asGroup.stages, firstError, asGroup).call()
```

Comment:
I might have missed it in my previous read through, but have you added validation and tests for empty groups?

Reply:
I have not. Good point.